### PR TITLE
array_api: Add an offset to `torch.eye`

### LINKED ISSUE #70910: `eye` should support other diagonals than the main one
The [array API specification stipulates](https://data-apis.org/array-api/latest/API_specification/creation_functions.html#eye-n-rows-n-cols-none-k-0-dtype-none-device-none) that the diagonal does not have to be the main one. This is controlled by the `k` parameter. Positive values correspond to upper diagonals whereas negative values correspond to lower diagonals. By default `k=0` which is the main diagonal and thus our current behavior.

### PR DESCRIPTION
**Pull request by HaoZeke**
Fixes #70910.
To-do status:
- [x] CPU implementation
- [x] Use a keyword argument for compatibility with `np.eye(4, k=-1)` 
- [x] Add tests
- [x] CUDA implementation
- ~[ ] Benchmark (?)~
- ~[ ] Cleanup history (?)~
- [x] Solicit reviews [Deferred]
   - [x] Internal
   - [x] External

### CODE REVIEW

**Comment by IvanYashchuk** [comment]
When a function signature is modified in `native_functions.yaml` then it requires adding it to the ALLOW_LIST in `test/forward_backward_compatibility/check_forward_backward_compatibility.py`. See the note in the file:
https://github.com/pytorch/pytorch/blob/c9ccb850abadc3fa1ef471e4630c677491032317/test/forward_backward_compatibility/check_forward_backward_compatibility.py#L12-L28

File: aten/src/ATen/native/TensorFactories.cpp
Comment (Lezcano) [inline aten/src/ATen/native/TensorFactories.cpp:432]: Need to first check that `m` and `k` actually have value, and if they dont, set them to their default value, right? Defining the arguments of the C++ funciton as `m_` or `k_` and then do
```c++
auto m = m_.has_value() ? m_.value() : n;
```
Then, you can do the `TORCH_CHECK`s as usual.

File: aten/src/ATen/native/TensorFactories.cpp
Comment (Lezcano) [inline aten/src/ATen/native/TensorFactories.cpp:434]: This is simply `k > -n && k < m`, right?

**Review by lezcano — CHANGES_REQUESTED** [review]
Currently the case when `m` and `k` are `None` should be failing in the CI, although this may be hidden simply by the other CI errors (there's a build error).

File: aten/src/ATen/native/TensorFactories.cpp
Comment (HaoZeke) [reply aten/src/ATen/native/TensorFactories.cpp:432]: I was wondering if it would be better to not assign to a new variable, but I see `TORCH_CHECK` isn't always the first thing in a function so, makes sense.

**Comment by HaoZeke** [comment]
> Currently the case when `m` and `k` are `None` should be failing in the CI, although this may be hidden simply by the other CI errors (there's a build error).

```cpp
  // m defaults to n (square matrix)
  // k defaults to 0 (main diagonal)
  c10::optional<int64_t> m_v{n}, k_v{0};
  if (m.has_value()) {
    m_v = m;
  }
  if (k.has_value()) {
    k_v = k;
  }
  auto tensor = at::empty({0}, options); // to be resized
  return at::eye_out(tensor, n, m_v, k_v);
```

This is means that subsequently, within `eye_out`, `m` and `k` are guaranteed to have values, so `torch.eye(3, None, None)` will work.

**Comment by Lezcano** [comment]
@shubhambhokare1 any pointers on how should the ONNX failures should be addressed?

**Comment by HaoZeke** [comment]
@Lezcano @IvanYashchuk This should now be ready for comments I believe :)

File: aten/src/ATen/native/LinearAlgebra.cpp
Comment (IvanYashchuk) [inline aten/src/ATen/native/LinearAlgebra.cpp:434]: Is this change necessary? It's fine to keep it, but I'd like to know whether `at::eye(self.size(-2), self.options())` still works and correct defaults are chosen.

File: torch/csrc/autograd/FunctionsManual.cpp
Comment (IvanYashchuk) [inline torch/csrc/autograd/FunctionsManual.cpp:1142]: Again, it's fine to keep this but it should also continue working without this change.

**Review by IvanYashchuk — APPROVED** [review]
Thank you for working on this!
I'm only curious whether `at::eye(int, options)` still work (I suspect it should).

**Comment by cpuhrsch** [comment]
@HaoZeke - This PR is currently marked as a draft. Is it ready for review?

File: aten/src/ATen/native/LinearAlgebra.cpp
Comment (HaoZeke) [reply aten/src/ATen/native/LinearAlgebra.cpp:434]: Unfortunately this is required (and is a BC breaking change). It isn't possible to disambiguate between `int, opt<int>, opt<int>, options` and any other combination. So the C++ API changes because of this PR. The Python usage is the same though.

Comment (HaoZeke) [reply aten/src/ATen/native/LinearAlgebra.cpp:434]: > @HaoZeke - This PR is currently marked as a draft. Is it ready for review?

Almost, will take it out of draft status for more discussion after resolving the remaining comments.

**Comment by HaoZeke** [comment]
> @HaoZeke - This PR is currently marked as a draft. Is it ready for review?

Almost, I'll remove it from the draft status after addressing the existing comments satisfactorily.

**Comment by HaoZeke** [comment]
@JackCaoG, I heard that you could perhaps suggest what might be going wrong for the XLA code-path?

**Comment by JackCaoG** [comment]
@HaoZeke yea xla needs to make a change to match this pr. Seems relatively easy. I will give you a ping when this is fixed.

File: torch/onnx/symbolic_opset9.py
Comment (garymm) [inline torch/onnx/symbolic_opset9.py:1905]: Since the number of args is fixed now, can you please change this from `*args` to `n, m, k, dtype, layout, device, pin_memory`?

File: torch/onnx/symbolic_opset9.py
Comment (garymm) [inline torch/onnx/symbolic_opset9.py:1908]:
```suggestion
    if is_m_none:
```

File: torch/onnx/symbolic_opset9.py
Comment (garymm) [inline torch/onnx/symbolic_opset9.py:1917]:
```suggestion
    if is_k_none:
```

File: torch/onnx/symbolic_opset9.py
Comment (garymm) [inline torch/onnx/symbolic_opset9.py:1907]: Is the check for `node().kind()` necessary if `type().kind() == "NoneType"`?
If it is not, then can we remove the check for `node().kind()`?

Same for the heck below on `is_k_none`

File: torch/onnx/symbolic_opset9.py
Comment (garymm) [inline torch/onnx/symbolic_opset9.py:1919]: Reduce nesting of the if, and leave the error case for the end:
```suggestion
    elif k.node().kind() == "onnx::Constant":
       ...
    else:
      raise RuntimeError(...)
```

File: torch/onnx/symbolic_opset9.py
Comment (garymm) [inline torch/onnx/symbolic_opset9.py:1921]: Please make the error message more informative. Maybe:
`Failed to export aten::eye to onnx::EyeLike because k is not a constant`

File: torch/onnx/symbolic_opset9.py
Comment (HaoZeke) [reply torch/onnx/symbolic_opset9.py:1959]: It does unfortunately seem to be required.

File: torch/onnx/symbolic_opset9.py
Comment (garymm) [reply torch/onnx/symbolic_opset9.py:1958]: not done?

**Review by garymm — CHANGES_REQUESTED** [review]

File: torch/onnx/symbolic_opset9.py
Comment (HaoZeke) [reply torch/onnx/symbolic_opset9.py]: Sorry, pushed a bad commit. Fixed.

**Review by garymm — APPROVED** [review]

**Review by IvanYashchuk — CHANGES_REQUESTED** [review]
As I understand the goal of this PR is to add to PyTorch two new overloads:
1. torch.eye(int n, int m, int k)
2. torch.eye(int n, int k=number)

Currently, we have:
- torch.eye(int n)
- torch.eye(int n, int m)

To add the overload (1) we can add one more entry to `native_functions.yaml` without removing or modifying what's already there:
```yaml
- func: eye.nmk(int n, int m, int k, *, options) -> Tensor
```
`nmk` here is the overload name that must be unique. Then to get the overload (2) we'd need to add:
```yaml
- func: eye.nk(int n, *, int k, options) -> Tensor
```
That way I think there would be no ambiguities and no breaking changes introduced. To get `eye(int n, int m, int k=number)` working then one more overload would need to be added. Yes, it's more tedious that way than having only one `eye(int n, int? m=None, int? k=None)` function, but there are constraints we have to work with.

However, the confusing part is the title of this PR, "numpy-compat", while the linked issue talks about the Array API compatibility, and the module label for this PR also indicates the Array API compatibility. NumPy's `eye` and Array API's `eye` are different. `k` is a positional argument in NumPy and it's a keyword-only argument in the Array API.

**Comment by rgommers** [comment]
> NumPy's `eye` and Array API's `eye` are different. `k` is a positional argument in NumPy and it's a keyword-only argument in the Array API.

`numpy.eye`'s `k` parameter is a keyword, not a positional argument. I.e., it can be used _both_ positionally and as `k=`. So the two are compatible, and numpy provides a superset of functionality of the array API standard.

Keeping the implementation in PyTorch as keyword-only should be fine, because that's the most idiomatic way to use the function (`eye(3, 4, 5)` is pretty hard to interpret). Although making it a regular keyword can also be done if it's easier (doesn't seem to be the case here, requires an extra overload).

File: aten/src/ATen/native/LinearAlgebra.cpp
Comment (HaoZeke) [reply aten/src/ATen/native/LinearAlgebra.cpp]: Changed to reflect a more BC respecting implementation as suggested :)

**Comment by HaoZeke** [comment]
@Lezcano and @IvanYashchuk I think this should finally be ready

**Comment by HaoZeke** [comment]
@Lezcano, gentle reminder :)

**Review by lezcano — APPROVED** [review]
LGTM. I don't know about the ONNX part of the code though, I'll let other people comment on those.

Not for this PR, but it looks like a simpler implementation of this operation that could work well for both CPU and CUDA is simply write it as we already do in CUDA
````c++
result.zero_();
result.diagonal(k).fill_(1.);
```

File: test/forward_backward_compatibility/check_forward_backward_compatibility.py
Comment (IvanYashchuk) [inline test/forward_backward_compatibility/check_forward_backward_compatibility.py:141]: I suspect that this change is not needed because now you're not modifying the existing signatures. Does the test break without this line?

**Review by IvanYashchuk — APPROVED** [review]
All my concerns are addressed. Thank you, Rohit, for working on this one, modifying signatures is a tricky business in ATen.

**Comment by HaoZeke** [comment]
> All my concerns are addressed. Thank you, Rohit, for working on this one, modifying signatures is a tricky business in ATen.

Thanks for helping this get into shape @IvanYashchuk @Lezcano and @garymm.

@mruberry perhaps this could be moved forward?

**Comment by HaoZeke** [comment]
@mruberry does this need any changes?

File: aten/src/ATen/native/cuda/TensorFactories.cu
Comment (mruberry) [inline aten/src/ATen/native/cuda/TensorFactories.cu:56]: This is probably redundant with checks that diagonal itself would do?

File: aten/src/ATen/native/cuda/TensorFactories.cu
Comment (mruberry) [inline aten/src/ATen/native/cuda/TensorFactories.cu:62]: It's probably a bug that this works for integer tensors (an inplace add of an integer tensor with a float scalar will, for example, throw an error). Let's just change `1.` to `1` to be more future proof?

File: aten/src/ATen/native/TensorFactories.cpp
Comment (mruberry) [inline aten/src/ATen/native/TensorFactories.cpp:500]: nit: let's fix this horrible formatting while we're here

File: aten/src/ATen/native/native_functions.yaml
Comment (mruberry) [inline aten/src/ATen/native/native_functions.yaml:2157]: in this signature `k` isn't optinal but required as a kwarg, right?

File: aten/src/ATen/native/native_functions.yaml
Comment (mruberry) [inline aten/src/ATen/native/native_functions.yaml:2171]: same question about this signature

**Review by mruberry — APPROVED** [review]
Cool! You can merge this when it's ready, @HaoZeke. I made some small inline comments for your review.

In the future we should probably OpInfo `torch.eye` (and more tensor creation ops!)

**Comment by rgommers** [comment]
> In the future we should probably OpInfo `torch.eye` (and more tensor creation ops!)

That reminded me of `TensorCreationOpInfo` by @pearu in gh-54187, which was never merged I think? Better to revive that than start over I'd think.

File: aten/src/ATen/native/native_functions.yaml
Comment (HaoZeke) [reply aten/src/ATen/native/native_functions.yaml:2157]: Yup, `k` isn't actually optional. Unfortunately:

```c++
 Tensor& eye_out_cuda(int64_t n, int64_t m, Tensor& result) {                                                                                                                           
   // the default value of `k` equals to 0                                                                                                                                              
   return at::native::eye_out_cuda(n, m, 0, result);                                                                                                                                    
 }                                                                                                                                                                                      
                                                                                                                                                                                        
Tensor& eye_out_cuda_nk(int64_t n, int64_t k, Tensor& result) {                                                                                                                        
   // the default value of `m` equals to `n`                                                                                                                                            
   int64_t k_v{k.has_value() ? k.value() : 0};                                                                                                                                          
   return at::native::eye_out_cuda(n, n, k_v, result);                                                                                                                                  
 }                                                                                                                                                                                      
         
```

These are ambiguous, since there's no way to know if the user passed in `k` or `m`. Actually the signature is:

```
eye(n_rows: int, n_cols: Optional[int] = None, /, *, k: int = 0, dtype: Optional[dtype] = None, device: Optional[device] = None) → array
```

Actually, the Array-API requires that `m` is to be optional, and `k` should be a keyword only argument. For BC reasons I ended up swapping the two; making `k` optional.

Comment (HaoZeke) [reply aten/src/ATen/native/native_functions.yaml:2157]: @IvanYashchuk does this seem like a reasonable approach?

Comment (IvanYashchuk) [reply aten/src/ATen/native/native_functions.yaml:2177]: There is an ambiguity with the "eye.m" overload at the C++ level as there's no concept of a keyword-only argument. Is it possible to add `eye.overload_name(int n, int? m=None, *, int k, ...)`?

File: aten/src/ATen/native/native_functions.yaml
Comment (IvanYashchuk) [inline aten/src/ATen/native/native_functions.yaml:2179]: Maybe add a comment here that this undocumented overload exists for NumPy compatibility, not Python Array API?

File: torch/_torch_docs.py
Comment (IvanYashchuk) [inline torch/_torch_docs.py:3728]: `k` is a keyword-only argument in the Array API.

File: torch/_torch_docs.py
Comment (IvanYashchuk) [inline torch/_torch_docs.py:3721]: `k` is a keyword-only argument in the Array API with default value 0.
```suggestion
eye(n, m=None, *, k=0, out=None, dtype=None, layout=torch.strided, device=None, requires_grad=False) -> Tensor
```

**Comment by Lezcano** [comment]
Let's land this one at https://github.com/pytorch/pytorch/pull/85145

**Pull request closed without merging**
